### ISSUE TITLE
Diablo II - dsoal-r435a and newer revisions produce an unknown software exception upon exiting the game

### ISSUE THREAD

**Issue by RavenMacDaddy** [kcat/dsoal#83]
Yes, even r435 works as intended, so the changes that were introduced in revision 435a specifically are apparently the culprit, which results in the error below.

![image](https://user-images.githubusercontent.com/67732686/235349932-4005a68e-249f-4e89-ad57-4845725b78f3.png)

For reference, the builds were obtained on this URL: http://vaporeon.io/hosted/dsoal-builds/

I first ruled out any other modifications and such to the game by using an installation that didn't have any, which also allowed me to notice the culprit in the first place.

I intended to use the latest version of DSOAL (at the time of writing, r441) with the game, so I placed the files in the path of the stock installation, upon which I started to experience the errors.

The next step was to trace where it began, so I went from the newest version and back. When I arrived at r435, the errors stopped.

**Comment by kcat** [comment]
Can you make a trace log with the latest version? If you set the `DSOAL_LOGLEVEL` env var to `3` and `DSOAL_LOGFILE` to a filename when you run the app, it'll write a log file to the named file that shows what it's doing leading up to the crash. It may work to create a .bat file (name something like `d2-dsoal-debug.bat`) and put this in it:
```
set DSOAL_LOGLEVEL=3
set DSOAL_LOGFILE=dsoal.log
<name>.exe
```
Put it in the same folder as the game executable, replace `<name>` with the game's main executable, and run the batch file instead of the game's main executable.

**Comment by RavenMacDaddy** [comment]
Here you go: [dsoal.log](https://github.com/kcat/dsoal/files/11128965/dsoal.log)

**Comment by kcat** [comment]
Is there anything particular you're doing when the crash happens? What do you do after starting the game and before it crashes?

**Comment by RavenMacDaddy** [comment]
Just starting and exiting the game will produce the same error pictured above consistently.

For reference, going in-game and then exiting doesn't result in a different error either.

**Comment by ThreeDeeJay** [comment]
Did you use the `-ns` parameter by any chance? According to https://github.com/fearedbliss/Cactus/blob/main/README-3D-SOUND.md#notes, it crashes (although with a different error) when using dsound.dll
You might wanna [get in touch](https://themovingcaravan.com/) with the Cactus devs cuz they might have more insight on the game's sound and possible issues.

**Comment by RavenMacDaddy** [comment]
Personally, I only dropped the DLLs into the directory for utilization, same as any previous versions of DSOAL.

The installation is simply the official files provided by Blizzard on their website through the corresponding offline installer, then updated to v1.14D using the Battle.net option in the game. 

Nevertheless, as the Cactus community seems to have a lot of knowledge on the matter, I did make a [post](https://themovingcaravan.com/viewtopic.php?t=207) over there to see if someone feels like assisting on solving the issue.

**Comment by ThreeDeeJay** [comment]
Hmm, you might also wanna try enabling logging in the game itself to see if there's any error/info outside DSOAL's scope.
While I'm not sure it'd provide relevant info, apparently [d2cs](https://manpages.ubuntu.com/manpages/bionic/man8/d2cs.8.html) has a `--debug` command and the Windows executable is available in [PvPGN](https://sourceforge.net/projects/pvpgn.berlios/files/latest/download).

**Comment by RavenMacDaddy** [comment]
The output seems strictly related to server functionality and not so much the game itself.

For reference, I can reproduce the problem by simply booting up the game and then immediately quitting - something related to sound is happening as the game closes, which combined with DSOAL r435a and onwards results in an application error.

With that said, I do appreciate any tips.

**Comment by ThreeDeeJay** [comment]
> Yes, even r435 works as intended, so the changes that were introduced in revision 435a specifically are apparently the culprit

I just got in touch with [Vaporeon](http://vaporeon.io/hosted/dsoal-builds/) and they think that r435 used unofficial builds of the latest OpenAL Soft release at the time (1.22.2), and r435**a** just updated it to 1.23.0.
![image](https://user-images.githubusercontent.com/71472458/235406797-a68c428a-411d-476f-bd50-8807437280be.png)
Although we're not sure of the exact commits, I was gonna suggest replacing dsoal-aldrv.dll with Win32 soft_oal.dll from commits between those releases to track down the exact OpenAL Soft commit that started causing issues (since dsound.dll/DSOAL is the same commit 435)
However, it seems most artifacts between 1.22.2 and 1.23.0 have already expired on both [Github actions](https://github.com/kcat/openal-soft/actions/runs/4050383701) and [Appveyor](https://ci.appveyor.com/project/ChrisRobinson/openal-soft/builds/46098171/job/5r251grj8a2qrjl6/artifacts) so perhaps @kcat or someone else happens to have backups of DLLs between those releases, or perhaps a different approach to take here 👀 

By the way, before going commit hunting, you should try [r444](http://vaporeon.io/hosted/dsoal-builds/dsoal-r444.7z) or [my automated builds](https://github.com/ThreeDeeJay/dsoal/releases/tag/latest) in case newer DSOAL/OpenAL Soft happened to fix this issue 🤔

**Comment by RavenMacDaddy** [comment]
Both you and @fearedbliss seemingly solved the case with later compiles that use v1.23.1 of OpenAL Soft.

As discussed in the topic over on his forums, v1.09b of the game doesn't crash with an application error, even with r444 - but 1.14d does.

Your builds don't, even with v1.14d of the game.

**Comment by RavenMacDaddy** [comment]
Closing this for now.
